### PR TITLE
Cok 2542 slice - Grammar/Modules/Attributes/Topics

### PR DESCRIPTION
**Pull request by davidcok**
<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

**Comment by alex-chew** [comment]
A general comment - the "(grammar)" text in each section title also appear in the table of contents, IMHO the TOC would look better without them.

**Comment by davidcok** [comment]
I agree. But unfortunately, the TOC is generated automatically from the section headings. And github markdown added a lot of vertical space to put the grammar link below the section headings. Having the links in the headings was the lesser problem, I thought. But I'm still looking for a way to accomplish both.

File: docs/DafnyRef/Grammar.md [hunk 1]
```diff
@@ -271,163 +168,115 @@ method m() {
...
+Dafny has a set of reserved words that may not
+be used as identifiers of user-defined entities.
+These are listed [here](#sec-g-tokens).
+
+In particular note that
+
+- `array`, `array2, `array3` , etc. are reserved words, but not `array1` or `array0`.
+These denote array types of given rank.
+- `array?`, `array2?, `array3?` , etc. are reserved words, but not `array1?` or `array0?`.
+These denote possibly-null array types of given rank.
```
Comment (alex-chew) [inline docs/DafnyRef/Grammar.md:186]:
```suggestion
- `array`, `array2`, `array3` , etc. are reserved words, but not `array1` or `array0`.
These denote array types of given rank.
- `array?`, `array2?`, `array3?` , etc. are reserved words, but not `array1?` or `array0?`.
These denote possibly-null array types of given rank.
```

File: docs/DafnyRef/Grammar.md [hunk 2]
```diff
@@ -271,163 +168,115 @@ method m() {
...
-A sequence of decimal digits, possibly interspersed with underscores for readability (but not beginning or ending with an underscore).
+A `digits` token is a sequence of decimal digits (`digit`), possibly interspersed with underscores for readability (but not beginning or ending with an underscore).
 Example: `1_234_567`.
-````grammar
-hexdigits = "0x" hexdigit {['_'] hexdigit}
-````
 
-A hexadecimal constant, possibly interspersed with underscores for readability (but not beginning or ending with an underscore).
+A `hexdigits` token denotes a hexadecimal constant, and is a sequence of hexadecimal digits (`hexdigit`)
+ possibly interspersed with underscores for readability (but not beginning or ending with an underscore).
```
Comment (alex-chew) [inline docs/DafnyRef/Grammar.md:206]: I think we should mention the `0x` prefix here.

**Review by alex-chew — COMMENTED** [review]
I've reviewed four of the five files so far; I'll come back to review `GrammarDetails.md`.

File: docs/DafnyRef/Grammar.md (hunk 2 above)
Comment (davidcok) [reply docs/DafnyRef/Grammar.md:206]: Agreed. Fixed.

**Review by alex-chew — DISMISSED** [review]

**Review by alex-chew — APPROVED** [review]

**Review by alex-chew — APPROVED** [review]

**Pull request merged**